ocrd/cli/process.py

```diff
@@ -12,30 +12,56 @@
 
 @click.command('process')
 @ocrd_cli_options
-@click.option('-T', '--ocrd-tool', multiple=True)
+@click.option('-T', '--ocrd-tool', multiple=True, type=click.Path(exists=True, dir_okay=False), help='register all executables from a ocrd-tool.json file')
 @click.argument('steps', nargs=-1)
-def process_cli(mets_url, **kwargs):
+def process_cli(mets, steps, **kwargs):
     """
     Execute OCR-D processors for a METS file directly.
+    
+    Run each of the STEPS executables one after another.
     """
     resolver = Resolver()
-    workspace = resolver.workspace_from_url(mets_url)
-
+    workspace = resolver.workspace_from_url(mets)
+    
     cmds = []
     for ocrd_tool_file in kwargs['ocrd_tool']:
         with codecs.open(ocrd_tool_file, encoding='utf-8') as f:
             obj = json.loads(f.read())
-            for tool in obj['tools']:
-                cmds.append(tool['binary'])
-
-    for cmd in kwargs['steps']:
+            for tool in obj['tools'].values():
+                cmds.append(tool['executable'])
+    
+    for cmd in steps:
         if cmd not in cmds:
             raise Exception("Tool not registered: '%s'" % cmd)
-
-    for cmd in kwargs['steps']:
-        run_cli(cmd, mets_url, resolver, workspace)
-
+    
+    # fixme: this really should be coming from workflow engine (giving steps, i/o fileGrps, parameter files)
+    # fixme: using the command line here is inefficient
+    # fixme: someone should go and check if the first input_file_grp actually exists
```

Comment:
Better yet: Check input requirements before each step and process success&output file group existence after every step.

Reply:
Yes. At the moment there is no error if a processor does not find any data. But that should be mocked IMO, and before the whole pipeline starts. Otherwise the user may need to undo the parts that were successful.